docs/faq.rst

```diff
@@ -50,27 +49,76 @@ Without `jit` the output is::
-is compiled and cached, and executed multiple times with different values of `x`, but
-with the same first value of `y`. 
+For ``jit``, the function is executed once using the Python interpreter, at which time the
+``Inside`` printing happens, and the first value of ``y`` is observed. Then the function
+is compiled and cached, and executed multiple times with different values of ``x``, but
+with the same first value of ```y``.
 
 Additional reading:
 
-  * [JAX - The Sharp Bits: Pure Functions](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#%F0%9F%94%AA-Pure-functions)
+  * `JAX - The Sharp Bits: Pure Functions <https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html#%F0%9F%94%AA-Pure-functions>`_.
+
+.. comment We refer to the anchor below in JAX error messages
+.. _value_must_be concrete:
+
+`Tracer value where concrete value is expected` error
+---------------------------------------------------------
+
+In the process of transforming functions JAX replaces some some function arguments with tracer values.
+You could see this if you use a ``print`` statement::
+
+  def func(x):
+    print(x)
+    return np.cos(x)
+
+  res = jax.jit(func)(0.)
+
+The above code does return the correct value ``1.`` but it also prints ``Traced<ShapedArray(float32[])>``
+for the value of ``x``. Normally, JAX handles these tracer values internally in a transparent way, e.g.,
+in the numeric JAX primitives that are used to implement the ``jax.numpy`` functions. This
+is why ``np.cos`` works in the example above.
+
+Some of the JAX primitives have special arguments for which tracers are not supported, typically arguments that
+control the shape and type of the result. For example,
+the 2nd and 3rd argument for :func:`jax.numpy.split` cannot contain tracers::
+
+  def split(arr, num_sections, axis):
+    return np.split(array, num_sections, axis)
+  split(np.zeros(2), 2, 0)  # works
+  jax.jit(split)(np.zeros(4), 2, 0)  # error
+
+will produce the following error::
+
+    TypeError: In jax.numpy.split argument 1: tracer value encountered where a concrete value is expected.
+    Use `functools.partial` or transformation parameters such as `static_argnums` for `jit` to preserve the concrete value during JAX tracing.
+    See `https://jax.readthedocs.io/en/latest/faq.html#value_must_be_concrete`.
+    Encountered value: Traced<ShapedArray(int32[], weak_type=True):JaxprTrace(level=-1/1)>
+
+This would happen with other JAX transformations, e.g., :func:`jax.grad`.
+We must change the way we use :func:`jax.jit` to ensure that the ``num_sections`` and ``axis`` arguments use
+their concrete values (``2`` and ``0`` respectively). For example, we can apply the transformation to a closure
+that encapsulates the arguments to be protected, either manually as below or by using ``functools.partial``::
+
+  jax.jit(lambda arr: split(arr, 2, 0))(np.zeros(4))
```

Comment:
This isn't a great pattern to encourage because it'll create a fresh function argument to `jit` on every invocation and thus prevent any compilation caching. Using `static_argnums` is better for caching, and is probably how we should recommend ensuring that some inputs to a `jit`ted function aren't traced.